command/agent/auth/cert/cert_test.go

```diff
@@ -0,0 +1,133 @@
+package cert
+
+import (
+	"context"
+	"os"
+	"path"
+	"reflect"
+	"testing"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/vault/api"
+	"github.com/hashicorp/vault/command/agent/auth"
+)
+
+func TestCertAuthMethod_Authenticate(t *testing.T) {
+	config := &auth.AuthConfig{
+		Logger:    hclog.NewNullLogger(),
+		MountPath: "cert-test",
+		Config: map[string]interface{}{
+			"name": "foo",
+		},
+	}
+
+	method, err := NewCertAuthMethod(config)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	client, err := api.NewClient(api.DefaultConfig())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	loginPath, _, authMap, err := method.Authenticate(context.Background(), client)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	expectedLoginPath := path.Join(config.MountPath, "/login")
+	if loginPath != expectedLoginPath {
+		t.Fatalf("mismatch on login path: got: %s, expected: %s", loginPath, expectedLoginPath)
+	}
+
+	expectedAuthMap := map[string]interface{}{
+		"name": config.Config["name"],
+	}
+	if !reflect.DeepEqual(authMap, expectedAuthMap) {
+		t.Fatalf("mismatch on login path:\ngot:\n\t%v\nexpected:\n\t%v", authMap, expectedAuthMap)
+	}
+}
+
+func TestCertAuthMethod_AuthClient_withoutCerts(t *testing.T) {
+	config := &auth.AuthConfig{
+		Logger:    hclog.NewNullLogger(),
+		MountPath: "cert-test",
+		Config: map[string]interface{}{
+			"name": "without-certs",
+		},
+	}
+
+	method, err := NewCertAuthMethod(config)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	client, err := api.NewClient(api.DefaultConfig())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	clientToUse, err := method.(auth.AuthMethodWithClient).AuthClient(client)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if !reflect.DeepEqual(client, clientToUse) {
```

Comment:
nit: This is a pretty good assertion to make, but might it be better to compare the actual pointer values? After all, if we are looking to test that we receive the same client back, the strongest assertion we can make is that it occupies the same area of memory which we know hasn't been garbage collected in the meantime. The `DeepEqual` assertion is more of a proxy for what we're trying to test, but we could get false negatives due to a deep copy.